runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,499 @@
+        return
+      end
+
+      -- Update the current tabstop to be the one containing the cursor.
+      local cursor_row, cursor_col = cursor_pos()
+      for tabstop_index, tabstops in pairs(M._session.tabstops) do
+        for _, tabstop in ipairs(tabstops) do
+          local range = tabstop:get_range()
+          if
+            (cursor_row > range[1] or (cursor_row == range[1] and cursor_col >= range[2]))
+            and (cursor_row < range[3] or (cursor_row == range[3] and cursor_col <= range[4]))
+          then
+            M._session.current_tabstop = tabstop
+            if tabstop_index ~= 0 then
+              return
+            end
+          end
+        end
+      end
+
+      -- The cursor is either not on a tabstop or we reached the end, so exit the session.
+      M.exit()
+    end,
+  })
+
+  vim.api.nvim_create_autocmd({ 'TextChanged', 'TextChangedI' }, {
+    group = group,
+    desc = 'Update active tabstops when buffer text changes',
+    buffer = M._session.bufnr,
+    callback = function()
+      -- Sync the tabstops in the current group.
+      local current_tabstop = M._session.current_tabstop
+      local current_text = current_tabstop:get_text()
+      for _, tabstop in ipairs(M._session.tabstops[current_tabstop.index]) do
+        if tabstop.extmark_id ~= current_tabstop.extmark_id then
+          tabstop:set_text(current_text)
+        end
+      end
+    end,
+  })
+end
+
+--- Clears the snippet autocommand group.
+---
+--- @private
+function M._clear_autocmds()
+  pcall(vim.api.nvim_clear_autocmds, { group = snippet_group, buffer = M._session.bufnr })
+end
+
+--- Expands the given snippet text.
+--- Refer to https://microsoft.github.io/language-server-protocol/specification/#snippet_syntax
+--- for the specification of valid input.
+---
+--- Tabstops are highlighted with hl-SnippetTabstop.
+---
+--- @param input string
+function M.expand(input)
+  local snippet = G.parse(input)
+  local snippet_text = {}
+
+  M._session = Session.new()
+
+  -- Keep track of tabstop nodes during expansion.
+  --- @type table<integer, Range4[]>
+  local tabstop_ranges = {}
+
+  --- @param index integer
+  --- @param placeholder string?
+  local function add_tabstop(index, placeholder)
+    tabstop_ranges[index] = tabstop_ranges[index] or {}
+    table.insert(tabstop_ranges[index], compute_tabstop_range(snippet_text, placeholder))
+  end
+
+  for _, child in ipairs(snippet.data.children) do
+    local type, data = child.type, child.data
+    if type == G.NodeType.Tabstop then
+      --- @cast data vim.snippet.TabstopData
+      add_tabstop(data.tabstop)
+    elseif type == G.NodeType.Placeholder then
+      --- @cast data vim.snippet.PlaceholderData
+      local value = tostring(data.value)
+      add_tabstop(data.tabstop, value)
+      append_to_snippet(snippet_text, value)
+    elseif type == G.NodeType.Choice then
+      --- @cast data vim.snippet.ChoiceData
+      append_to_snippet(snippet_text, data.values[1])
```

Comment:
this looks to me like we just insert the first value of a choice. Is it planned to support actually allowing choices in a future pr or is this out of scope for core?

Reply:
Please read the roadmap I outlined in this PR's description.